Stripe/Payments/STPPaymentHandler.m

```diff
@@ -344,101 +367,99 @@ - (BOOL)handleURLCallback:(NSURL * __unused)url {
     [[STPURLCallbackHandler shared] unregisterListener:self];
     [[_currentAction.authenticationContext authenticationPresentingViewController] dismissViewControllerAnimated:YES completion:nil];
-    [self _retrieveAndCheckPaymentIntentForCurrentAction];
+    [self _retrieveAndCheckIntentForCurrentAction];
     return YES;
 }
 
 #pragma mark - STPChallengeStatusReceiver
 
 - (void)transaction:(__unused STDSTransaction *)transaction didCompleteChallengeWithCompletionEvent:(STDSCompletionEvent *)completionEvent {
-    STPPaymentHandlerActionCompletionBlock completion = _currentAction.completion;
-    NSAssert(completion != nil, @"Shouldn't have a nil completion block at this point.");
-    if (completion == nil) {
-        return;
-    }
     NSString *transactionStatus = completionEvent.transactionStatus;
+    __weak __typeof(self) weakSelf = self;
     if ([transactionStatus isEqualToString:@"Y"]) {
         [self _markChallengeCompletedWithCompletion:^(BOOL markedCompleted, NSError * _Nullable error) {
-            completion(markedCompleted ? STPPaymentHandlerActionStatusSucceeded : STPPaymentHandlerActionStatusFailed, self->_currentAction.paymentIntent, error);
+            __typeof(self) strongSelf = weakSelf;
+            [strongSelf->_currentAction completeWithStatus:markedCompleted ? STPPaymentHandlerActionStatusSucceeded : STPPaymentHandlerActionStatusFailed error:error];
         }];
 
     } else {
         // going to ignore the rest of the status types because they provide more detail than we require
         [self _markChallengeCompletedWithCompletion:^(__unused BOOL markedCompleted, __unused NSError * _Nullable error) {
-            completion(STPPaymentHandlerActionStatusFailed, self->_currentAction.paymentIntent, [self _errorForCode:STPPaymentHandlerNotAuthenticatedErrorCode userInfo:@{@"transaction_status": transactionStatus}]);
+            __typeof(self) strongSelf = weakSelf;
+            [strongSelf->_currentAction completeWithStatus:STPPaymentHandlerActionStatusFailed error:[self _errorForCode:STPPaymentHandlerNotAuthenticatedErrorCode userInfo:@{@"transaction_status": transactionStatus}]];
         }];
     }
 }
 
 - (void)transactionDidCancel:(__unused STDSTransaction *)transaction {
-    STPPaymentHandlerActionCompletionBlock completion = _currentAction.completion;
-    NSAssert(completion != nil, @"Shouldn't have a nil completion block at this point.");
-    if (completion == nil) {
-        return;
-    }
+    __weak __typeof(self) weakSelf = self;
     [self _markChallengeCompletedWithCompletion:^(__unused BOOL markedCompleted, __unused NSError * _Nullable error) {
-        completion(STPPaymentHandlerActionStatusCanceled, self->_currentAction.paymentIntent, nil);
+        __typeof(self) strongSelf = weakSelf;
+        [strongSelf->_currentAction completeWithStatus:STPPaymentHandlerActionStatusCanceled error:nil];
```

Comment:
this will crash if `strongSelf` is nil.

Reply:
Ahhh I'll have to add checks everywhere, thanks.